### PR TITLE
feat: Removing stack trace info in production env

### PR DESCRIPTION
**Pull request by MomentQYC**
## What

Remove stack trace information from production errors, and replace a small number of unnecessary errors with other types of output.

## Why

This is what `instance/queue/` currently looks like when accessed without logging in (e.g. `misskey.io/queue`)
image
Compare the changed look
image
Many times in production environments stack trace information is unnecessary and a security risk.
For a small number of output changes, I think an average user would prefer a program that only outputs logs as opposed to a program that reports errors when a login action is required.

## Checklist
- [x] Read the [contribution guide](https://github.com/misskey-dev/misskey/blob/develop/CONTRIBUTING.md)
- [x] Test working in a local environment
- [ ] (If needed) Add story of storybook
- [ ] (If needed) Update CHANGELOG.md
- [ ] (If possible) Add tests

### CODE REVIEW

**Comment by tamaina** [comment]
I think it's better to use ApiError in API execution service

**Comment by tamaina** [comment]
There should be no need to change the frontend

**Comment by tamaina** [comment]
I don't know why the test is failing

**Comment by tamaina** [comment]
Somehow Jest can't use `./error.js`

**Comment by MomentQYC** [comment]
> There should be no need to change the frontend

The main consideration is to facilitate centralized management and reduce unnecessary errors.

**Comment by MomentQYC** [comment]
> > There should be no need to change the frontend
> 
> The main consideration is to facilitate centralized management and reduce unnecessary errors.

For Example, https://github.com/misskey-dev/misskey/blob/827616f63094997bd44458b8c072a50da176af32/packages/frontend/src/scripts/please-login.ts#L25C3-L25C3

**Comment by tamaina** [comment]
> facilitate centralized management and reduce unnecessary errors.

I don't want them on the frontend.

**Comment by tamaina** [comment]
I think it is not preferable to remove stack traces or change Errors to console.log on the front end.

**Comment by MomentQYC** [comment]
> I think it is not preferable to remove stack traces or change Errors to console.log on the front end.

I have modified it to use `console.log` in the production environment, while still throwing an error in other environments. Is this feasible?

**Comment by tamaina** [comment]
Do not rewrite `throw Error`s in frontend. They are there to stop calling(upstream) functions, for example.

**Comment by MomentQYC** [comment]
> Do not rewrite throw Errors in frontend. They are there to stop calling(upstream) functions, for example.

 OK, I will remove all frontend changes first.

**Comment by tamaina** [comment]
This PR should not include frontend changes.

**Comment by saschanaz** [comment]
I'm not sure we want `throw ErrorHandling`, because errors can be from anywhere, from libraries and native JS functions. What we probably want is to customize Fastify error handler: https://fastify.dev/docs/latest/Reference/Server/#seterrorhandler

**Review by saschanaz — CHANGES_REQUESTED** [review]
As above.

**Comment by MomentQYC** [comment]
> I'm not sure we want `throw ErrorHandling`, because errors can be from anywhere, from libraries and native JS functions. What we probably want is to customize Fastify error handler: https://fastify.dev/docs/latest/Reference/Server/#seterrorhandler

I would like to address the existing problems first and then focus on improving error-related problems in the future. Since I'm not very familiar with TypeScript, it might take some time for me to make changes to the code (if necessary).

**Comment by saschanaz** [comment]
>I would like to address the existing problems first 

I highly suspect "addressing the existing problems" requires more code than writing proper error handler, which should require change in only one file.

**Comment by MomentQYC** [comment]
> As above.

Like now?
Defined errors still remove tracking information as before, undefined errors are processed further.

**Comment by saschanaz** [comment]
There's probably no need to do setErrorHandler in every createServer, you should be able to do it only once in ServerService.

**Comment by saschanaz** [comment]
Given that this PR now sets an error handler that drops stack info, the changes in other files are not needed, right?

**Comment by MomentQYC** [comment]
> Given that this PR now sets an error handler that drops stack info, the changes in other files are not needed, right?

My original intention was to minimize stack trace information in production environments as much as possible, and in particular, as much as possible, not to expose the user to stack trace information especially from the back-end part.
So I think especially in scenarios like the one in the very first screenshot, targeted removal of stack trace information is necessary, other than that it's optional.

**Comment by saschanaz** [comment]
You mean there will be any situation that the stack will still be exposed even with the error handler?

**Comment by MomentQYC** [comment]
> You mean there will be any situation that the stack will still be exposed even with the error handler?

When I tested along the lines of the code here in the production environment, I still found the stack trace.
https://github.com/misskey-dev/misskey/blob/ab58b651f79e182c20a238f9c07f0c0006a58599/packages/backend/src/server/web/ClientServerService.ts#L148C5-L151C6

**Comment by saschanaz** [comment]
Hmm, that's specific to Misskey's bull-board usage where [the library registers its own error handler](https://github.com/felixmosh/bull-board/blob/e0c176a92eb612d73e7c402e3fa7ba59c344920e/packages/fastify/src/FastifyAdapter.ts#L151-L154) (and thus yours is being ignored). It seems reasonable to not throw an error but explicitly send an object instead in https://github.com/misskey-dev/misskey/blob/ab58b651f79e182c20a238f9c07f0c0006a58599/packages/backend/src/server/web/ClientServerService.ts#L148C5-L151C6, as [their example does](https://github.com/felixmosh/bull-board/blob/e0c176a92eb612d73e7c402e3fa7ba59c344920e/examples/with-fastify-auth/cookieAuth.js#L66), and later investigate this bull-board issue separately.

For other cases the error handler should be sufficient.

**Comment by MomentQYC** [comment]
This is where it's most likely to be seen by the user, so here's where I'd like to apply additional error handling (removing the stack trace).
This has been changed to `reply.code().send()` in the latest changes.
Like this:
image

**Comment by syuilo** [comment]
やる場合、new Errorの使用を禁止する何らかの機構(lintとかgithub actionsとか)が無いとこのPRの目的を維持するのは難しそう

**Comment by saschanaz** [comment]
> This is where it's most likely to be seen by the user, so here's where I'd like to apply additional error handling (removing the stack trace).
> This has been changed to `reply.code().send()` in the latest changes.
> Like this:
> image

What I'm saying is that it's doable with error handler, it's the sole purpose of error handler anyway. The reason it's not working for bull-board one is explained above, and I suggest using send() only for that exception for now and reverting others. And maybe add some test by adding some test-only route.

**Comment by MomentQYC** [comment]
> reverting others

Do you mean like now?

File: packages/backend/src/error.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+/*
+ * SPDX-FileCopyrightText: tamaina and other misskey contributors
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+// Never use `./error.js` because jest can't use it.
```
Comment (saschanaz) [inline packages/backend/src/error.ts]: I guess this file is redundant?

File: packages/backend/src/misc/fastify-reply-error.ts [hunk 2]
```diff
@@ -12,5 +12,9 @@ export class FastifyReplyError extends Error {
 		super(message);
 		this.message = message;
 		this.statusCode = statusCode;
+
+		if (process.env.NODE_ENV === 'production') {
+			Object.defineProperty(this, 'stack', { value: '' });
```
Comment (saschanaz) [inline packages/backend/src/misc/fastify-reply-error.ts]: I guess `this.stack = undefined` is more consistent.

File: packages/backend/src/server/oauth/OAuth2ProviderService.ts [hunk 3]
```diff
@@ -353,6 +353,7 @@ export class OAuth2ProviderService {
 	public async createServer(fastify: FastifyInstance): Promise<void> {
 		// https://datatracker.ietf.org/doc/html/rfc8414.html
 		// https://indieauth.spec.indieweb.org/#indieauth-server-metadata
+
```
Comment (saschanaz) [inline packages/backend/src/server/oauth/OAuth2ProviderService.ts]: Redundant

File: packages/backend/src/misc/error.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+		error.stack = undefined;
+	}
+	if (reply) {
+		reply.code(statusCode ?? 500);
+	}
+	return error;
+}
+
+export function ErrorHandler(error: Error, request: FastifyRequest, reply: FastifyReply): void {
+	throw ErrorHandling(error.message, reply);
```
Comment (saschanaz) [inline packages/backend/src/misc/error.ts]: Merge these two functions?

File: packages/backend/src/misc/error.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ */
+import { FastifyReply, FastifyRequest } from 'fastify';
+
+export function ErrorHandling(message: string, reply?: FastifyReply, statusCode?: number): Error {
+	const error = new Error(message);
+	if (process.env.NODE_ENV === 'production') {
+		error.stack = undefined;
+	}
+	if (reply) {
+		reply.code(statusCode ?? 500);
```
Comment (saschanaz) [inline packages/backend/src/misc/error.ts]: You can skip this if you do `reply.send(error)`, the send function will pass the error to the default error handler where it'll process the status code.

**Review by saschanaz — CHANGES_REQUESTED** [review]

File: packages/backend/src/misc/error.ts (hunk 4 above)
Comment (MomentQYC) [reply packages/backend/src/misc/error.ts]: The lack of merging here is in consideration of possible separate application scenarios.

Comment (saschanaz) [reply packages/backend/src/misc/error.ts]: We can split it again if it's needed later of course, but in the current state I see no reason for that.

**Comment by saschanaz** [comment]
Hmm, I locally tested what Fastify does by adding a random error, but it seems Fastify already does not expose error stack information by default at all. The [`details` field you saw in `/queue` was from bull-board custom error handler](https://github.com/felixmosh/bull-board/blob/e0c176a92eb612d73e7c402e3fa7ba59c344920e/packages/api/src/handlers/error.ts#L9) and nothing to do with Fastify default behavior, which only adds `statusCode`, `error`, and `message`.

tl;dr: we don't need setErrorHandler at all, this PR should be sufficient only with changes on `ClientServerService.ts`.

**Comment by MomentQYC** [comment]
> we don't need setErrorHandler at all

Reverted

**Review by saschanaz — APPROVED** [review]
👍

**Comment by syuilo** [comment]
👍🏻
